### PR TITLE
[SPARK-45137][CONNECT] Support map/array parameters in parameterized `sql()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to extend the proto protocol, and add two more fields to `SqlCommand`: `arguments` and `pos_arguments`. This should allow to construct arrays/maps/structs by the `map()`, `array()`, `struct()`, `map_from_arrays()` and `map_from_entries()`, and pass them to `sql()` as arguments.

### Why are the changes needed?
To fix the issue demonstrated by the example:
```scala
spark.sql("select element_at(?, 1)", Array(array(lit(1)))).collect()
java.lang.UnsupportedOperationException: literal unresolved_function {
[info]   function_name: "array"
[info]   arguments {
[info]     literal {
[info]       integer: 1
[info]     }
[info]   }
[info] }
[info]  not supported (yet).
```

Note: The PR https://github.com/apache/spark/pull/42752 supported constructors, and this PR extends the feature on Scala connect client.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By new checks:
```
$ build/sbt -Phive-2.3 -Phive-thriftserver "test:testOnly *.ClientE2ETestSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 1]
```diff
@@ -65,6 +65,12 @@ message SqlCommand {
 
   // (Optional) A sequence of literal expressions for positional parameters in the SQL query text.
   repeated Expression.Literal pos_args = 3;
+
+  // (Optional) A map of parameter names to expressions.
+  map<string, Expression> arguments = 4;
```
Comment (cloud-fan) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:70]: `named_arguments`?

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 2]
```diff
@@ -65,6 +65,12 @@ message SqlCommand {
...
+
+  // (Optional) A sequence of expressions for positional parameters in the SQL query text.
```
Comment (cloud-fan) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:72]: shall we doc that this can't coexist with `named_arguments`?

Comment (cloud-fan) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:72]: shall we also mention that the `args` and `pos_args` are deprecated?

Comment (MaxGekk) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:72]: @grundprinzip @HyukjinKwon Is there any way to deprecate some fields?

Comment (MaxGekk) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:72]: Would this be enough `[deprecated=true]`?

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @HyukjinKwon and @cloud-fan for review.

**Pull request closed without merging**